### ISSUE TITLE
Performance when #bpe is increased.

### ISSUE THREAD

**Issue by jubang0219** [espnet/espnet#2154]
Hi there,

I'm investigating the performance of Korean speech data (960 hours) according to various #bpe.
(Our data consists approximately 2,300 characters (=syllables in Korean).)

So, I'm training the transformer model by increasing #bpe to 2.3k, 5k, 8k, 10k, and 12k.
(The same parameter setting, only #bpe is different.)

But, I observed that training is performing strangely, when #bpe is 10k and 12k.
-> In the Acc & Loss figure below, I'm not sure why the acc is dropping in #bpe 10k and 12k.

Can I think that is a problem of **data shortage** due to too large #bpe?

Thanks.

Bang

---- 
**# stage 2 in run.sh**
[unk]: oov symbol present in transcription.
bpemode: unigram
```
    spm_train --input=data/lang_char/input.txt --user_defined_symbols='[unk]' --character_coverage=1.0 \
              --vocab_size=${nbpe} --model_type=${bpemode} --model_prefix=${bpemodel} --input_sentence_size=100000000 
    spm_encode --model=${bpemodel}.model --output_format=piece < data/lang_char/input.txt | tr ' ' '\n' | sort | uniq | awk '{print $0 " " NR+1}' >> ${dict}
```
----
**# Training configuration with 4 gpus (1080Ti)**
accum-grad: 5
adim: 512
aheads: 8
backend: pytorch
batch-bins: 12000000
dlayers: 6
dropout-rate: 0.1
dunits: 2048
elayers: 12
epochs: 120
eunits: 2048
grad-clip: 5
lsm-weight: 0.1
model-module: espnet.nets.pytorch_backend.e2e_asr_transformer:E2E
mtlalpha: 0.3
opt: noam
patience: 0
sortagrad: 0
transformer-attn-dropout-rate: 0.0
transformer-init: pytorch
transformer-input-layer: conv2d
transformer-length-normalized-loss: false
transformer-lr: 10.0
**transformer-warmup-steps: 25000**

----
**# Training log for #bpe 10k**
 [J     total [###...............................................]  7.27%
this epoch [####################################..............] 72.16%
     24900 iter, 8 epoch / 120 epochs
    0.3613 iters/sec. Estimated time to finish: 10 days, 4:14:59.077851.
 [4A [J8           25000       30.685      41.791         25.9252                                                                                  0.705392                                                                70064.8       
 [J     total [###...............................................]  7.30%
this epoch [#####################################.............] 75.67%
     **25000 iter**, 8 epoch / 120 epochs
   0.36135 iters/sec. Estimated time to finish: 10 days, 4:08:17.967922.
 [4A [J8           25100       30.7156     41.9505        25.9006                                                                                  0.70575                                                                 70333.6       
 [J     total [###...............................................]  7.33%
this epoch [#######################################...........] 79.17%
     25100 iter, 8 epoch / 120 epochs
    0.3614 iters/sec. Estimated time to finish: 10 days, 4:01:46.310691.

----
**# Training log for #bpe 12k**
 [4A [J7           24900       52.6172     67.7491        46.1322                                                                                  0.576962                                                                69109         
 [J     total [###...............................................]  6.41%
this epoch [##################################................] 69.24%
     24900 iter, 7 epoch / 120 epochs
   0.36198 iters/sec. Estimated time to finish: 11 days, 14:58:12.494850.
 [4A [J7           25000       54.152      70.0712        47.3295                                                                                  0.578835                                                                69378.9       
 [J     total [###...............................................]  6.44%
this epoch [####################################..............] 72.33%
     **25000 iter**, 7 epoch / 120 epochs
   0.36204 iters/sec. Estimated time to finish: 11 days, 14:50:58.769736.
 [4A [J7           25100       48.277      63.545         41.7336                                                                                  0.566857                                                                69650.2       
 [J     total [###...............................................]  6.46%
this epoch [#####################################.............] 75.42%
     25100 iter, 7 epoch / 120 epochs
   0.36203 iters/sec. Estimated time to finish: 11 days, 14:46:48.418473.

----
**# Acc & Loss**
![acc](https://user-images.githubusercontent.com/21096732/86568536-f93a1b80-bfa7-11ea-8af0-03ae6907bdbf.png)
![loss](https://user-images.githubusercontent.com/21096732/86564629-80d05c00-bfa1-11ea-8417-2d138fcdb9fd.png)

**Comment by sw005320** [comment]
I think such strange behaviors can be fixed by tuning either of the following parameters
```
transformer-lr: 10.0
transformer-warmup-steps: 25000
```
This would not be due to the small amount of training data.

**Comment by kan-bayashi** [comment]
I think one of the reasons of the strange behavior is `batch-bins: 12000000`.
If you change the BPE size while fixing batch-bins, the actual batch size will be small.

**Comment by sw005320** [comment]
You're right.
Now, I understood this strange behavior well.
Instead of tuning the above hyperparameter, please tune batch-bins (make it higher).
Check the documentation about batch-bins.

**Comment by jubang0219** [comment]
Thank you for your comments.

I understood that the batch-bins is related to the output dimension (#BPE).

_# Batch bins :: Dynamic batch size depending on size of samples
batch-bins = avg batch-size * (mean_out * dim_out + mean_in * dim_in)_

Based on your comments, I will increase the "batch-bins" from 12M to 18M.
And, if the OOM problem occurs, I will increase #gpus from 4 to 6. (<- is this okay?
As far as I understand, increasing #gpu is the same as increasing the actual batch size.)

Thanks.

Bang

**Comment by sw005320** [comment]
Yes, you're right.
But probably, you can get some normal behaviors when you try to increase the batch size to the limit of the GPU memory.
Then if you still encounter the strange learning curve behavior, tuning the following hyperparameters is another option.
```
transformer-lr: 10.0
transformer-warmup-steps: 25000
```

**Comment by JiayiFu** [comment]
Maybe change the batch mode from bins to sequences or frames can eliminate the impact of BPE number on batch size.

**Comment by JiayiFu** [comment]
> I think such strange behaviors can be fixed by tuning either of the following parameters
> 
> ```
> transformer-lr: 10.0
> transformer-warmup-steps: 25000
> ```
> 
> This would not be due to the small amount of training data.

Hi @sw005320, I found the librispeech recipe also uses the same transformer-lr and transformer-warmup-steps values, 10.0 and 25k. When there are more or fewer data, how to tune these hyperparameters?

**Comment by sw005320** [comment]
You can decrease `transformer-lr`.

**Comment by jubang0219** [comment]
Thank you for your advice.

I got the normal behaviors when I 1) increased the batch-bins, 2) decreased the learning-rate, or 3) increased the warmup-steps.

![acc_new](https://user-images.githubusercontent.com/21096732/86900916-e90c7280-c146-11ea-8987-071bc86d769a.png)

Thanks.

Bang

**Comment by sw005320** [comment]
Thanks!
This is a very good TIPS!

**Comment by AdolfVonKleist** [comment]
This is not directly related to espnet, but the topics in this thread are also addressed in a general fashion in this paper:

Training Tips for the Transformer Model
* https://arxiv.org/abs/1804.00247

which is definitely interesting, and worth reading IMO if you have not come across it yet.
